src/server/services/wfs/qgswfs.cpp

```diff
@@ -71,7 +72,15 @@ namespace QgsWfs
 
         if ( QSTR_COMPARE( req, "GetCapabilities" ) )
         {
-          writeGetCapabilities( mServerIface, project, versionString, request, response );
+          // Supports WFS 1.0.0
+          if ( QSTR_COMPARE( versionString, "1.0.0" ) )
+          {
+            v1_0_0::writeGetCapabilities( mServerIface, project, versionString, request, response );
```

Comment:
I'm not a big fan of this notation : `v1_0_0::`. 

Moreover, I've the feeling that most of the code is duplicated between `qgswfsgetcapabilities.cpp` and `qgswfsgetcapabilities_1_0_0.cpp`. And it's not homogeneous with other services (for example, we don't have a `qgswmsrenderer.cpp` and a `qgswmsrenderer_1_3_0.cpp`). 

No? Or am I missing something?

Reply:
The WFS capabilities documents are not the same at all and it will be easiest to enhance or maintain with separate files. For the other request like DescribeFeature or GetFeature, we reused the same code.

Reply:
It's true that it's not as similar as I thought... However, I think we could factor a little more.

For example, `writeGetCapabilities` is identical and there's some redundant content in `createGetCapabilitiesDocument` (WFS_Capabilities, ogc:SpatialOperators) or `getFeatureTypeListElement`.

That being said, your comment about maintainability is justified...

@nyalldawson @m-kuhn What do you think about adding files for specific service versions (like those introduced in this PR : `qgswfsgetcapabilities.cpp` and `qgswfsgetcapabilities_1_0_0.cpp`) ?

Reply:
FWIW, owslib made the same thing as you (separate versions in files) : https://github.com/geopython/OWSLib/tree/master/owslib/feature

So why not, but in this case, maybe we should have a `qgswfsgetcapabilities_common.x`, `qgswfsgetcapabilities_100.x` and `qgswfsgetcapabilities_110.x` files to be more explicit?

Otherwise (I'm just thinking out loud, so it may probably be a bad idea...), maybe a composite/visitor pattern could be used:
- 1 node of the tree for 1 node of the document 
- one or several validity versions for each node (the same node may be valid for 1.0.0 or 1.1.0)
- a visitor collects only nodes for a specific version
- when we want to add a new version, we just have to add some node and retrieve existing nodes

Reply:
Hi @pblottiere
 
There is several ways to add new versions:
- Handle several versions in one module (all in one, only one default version registred but assumed to  handle everything, or register multiple versions in the same module.)
- Add new versions with different modules python or c++ (register them with there proper versions)

In the present the case we use a different namespace juste to separates units specific to a given version: I do not care about the name of the namespace as long it is meaningful in the contexte - IMO, the advantage of using namespace over is that you don't have to rename all your fonctions/objects. 

IMHO, we should consider the current implementation as a easy way to implement quickly  the 1_1_0 without taking the chance to break actual implementation with more or less heavy refactoring: we will have time to factorize things when all the stuff will be stabilized a little more.

Reply:
Hi @dmarteau 
I don't have a strong opinion against the use of a namespaces here as long as it does not introduce redundancy in the code.
And I would also explicit versions: it is a bit strange to have qgswfsgetcapabilities_1_0_0 and qgswfsgetcapabilities without suffix (which version ?), no ?

Reply:
@mhugo it's because 1.1.0 will be the default version instead of 1.0.0 which become a special version.

Reply:
@mhugo, @rldhont  there is no points to make a debat about it, I really do not care about it as long as naming is not ambiguous.